python/grass/tools/session_tools.py

```diff
@@ -17,10 +17,82 @@
 import os
 
 import grass.script as gs
+from grass.exceptions import CalledModuleError
 
 from .support import ParameterConverter, ToolFunctionResolver, ToolResult
 
 
+class ToolError(CalledModuleError):
+    """Raised when a tool run ends with error (typically a non-zero return code)
+
+    Inherits from *subprocess.CalledProcessError* to make it easy to transition from
+    or combine with code which is using the *subprocess* package.
+    Inherits from CalledModuleError to make it easy to transition code with except
+    statements around *grass.script.run_command*-style tool calls, but new code should
+    not rely on that.
+    """
+
+    def __init__(self, tool, cmd, returncode, errors=None):
```

Comment:
Would it make sense to add typing annotations here? For CalledModuleError, I had a bit of difficulty making sure what was accepted and expected, and it is used in the wild so harder to make a change if there was a problem. 

Here it is completely and you still know what are your design intents

Reply:
Good point. Myself, maybe others too, tried to accommodate too much all possible use cases with CalledModuleError (grass.script and grass.pygrass) and never really spec'd out what it should do. ToolError is more narrow in that sense.